File: tests/mocks/pyspark/modular_app/data_output_module.py
"""
data_output_module.py
Handles all data writing operations to various sinks
"""

from pyspark.sql import DataFrame
from typing import Optional


class DataOutputModule:
    """Module responsible for writing data to various sinks"""
    
    def __init__(self, spark, env: str = "prod"):
        self.spark = spark
        self.env = env
    
    def write_enriched_transactions(self, df: DataFrame, run_date: str) -> None:
        """Write enriched transactions to Hive Gold layer"""
        table_name = f"{self.env}_gold.enriched_transactions"
        print(f"Writing enriched transactions to: {table_name}")
        
        (df.write
         .mode("append")
         .partitionBy("run_date")
         .format("parquet")
         .saveAsTable(table_name))
    
    def write_customer_metrics(self, df: DataFrame, run_date: str, 
                               base_path: str = "/data/gold") -> None:
        """Write customer metrics to HDFS"""
        path = f"{base_path}/customer_metrics/date={run_date}"
        print(f"Writing customer metrics to: {path}")
        
        (df.write
         .mode("overwrite")
         .partitionBy("region")
         .parquet(path))
    
    def write_fraud_alerts(self, df: DataFrame, 
                          jdbc_url: str = "jdbc:postgresql://prod-pg:5432/alerts") -> None:
        """Write fraud alerts to PostgreSQL"""
        print(f"Writing fraud alerts to: {jdbc_url}")
        
        (df.write
         .mode("append")
         .format("jdbc")
         .option("url", jdbc_url)
         .option("dbtable", "fraud_alerts")
         .option("user", "etl_user")
         .option("password", "${PG_PASSWORD}")
         .option("driver", "org.postgresql.Driver")
         .save())
    
    def write_product_summary(self, df: DataFrame) -> None:
        """Write product summary to Hive reporting layer"""
        table_name = f"{self.env}_reports.product_daily_summary"
        print(f"Writing product summary to: {table_name}")
        
        (df.write
         .mode("overwrite")
         .format("orc")
         .option("compression", "snappy")
         .saveAsTable(table_name))
    
    def write_inventory_status(self, df: DataFrame, run_date: str) -> None:
        """Write inventory status to Hive warehouse table"""
        table_name = f"{self.env}_warehouse.inventory_status"
        print(f"Writing inventory status to: {table_name}")
        
        (df.write
         .mode("overwrite")
         .insertInto(table_name))
    
    def write_events_to_kafka(self, df: DataFrame, topic: str = "processed-events") -> None:
        """Write processed events to Kafka"""
        print(f"Writing processed events to Kafka topic: {topic}")
        
        (df.selectExpr("CAST(customer_id AS STRING) AS key", "to_json(struct(*)) AS value")
         .writeStream
         .format("kafka")
         .option("kafka.bootstrap.servers", "kafka-broker:9092")
         .option("topic", topic)
         .option("checkpointLocation", f"/tmp/checkpoints/{topic}")
         .trigger(processingTime="30 seconds")
         .start())
    
    def write_to_delta(self, df: DataFrame, table_name: str, 
                      base_path: str = "/data/delta") -> None:
        """Write data to Delta Lake"""
        path = f"{base_path}/{table_name}"
        print(f"Writing to Delta Lake: {path}")
        
        (df.write
         .format("delta")
         .mode("overwrite")
         .option("mergeSchema", "true")
         .save(path))
    
    def export_to_s3(self, df: DataFrame, report_name: str, run_date: str,
                    s3_bucket: str = "s3a://company-reports") -> None:
        """Export reports to S3 as CSV"""
        path = f"{s3_bucket}/daily/{report_name}/date={run_date}"
        print(f"Exporting report to S3: {path}")
        
        (df.write
         .mode("overwrite")
         .option("header", "true")
         .option("compression", "gzip")
         .csv(path))
    
    def write_customer_segments(self, df: DataFrame, run_date: str) -> None:
        """Write customer segments to Hive analytics table"""
        table_name = f"{self.env}_analytics.customer_segments"
        print(f"Writing customer segments to: {table_name}")
        
        (df.write
         .mode("append")
         .partitionBy("run_date")
         .format("parquet")
         .saveAsTable(table_name))
    
    def write_daily_snapshot(self, df: DataFrame, run_date: str,
                            base_path: str = "/data/snapshots") -> None:
        """Write daily snapshot to HDFS"""
        path = f"{base_path}/daily_snapshot/date={run_date}"
        print(f"Writing daily snapshot to: {path}")
        
        (df.write
         .mode("overwrite")
         .parquet(path))
    
    def write_to_redshift(self, df: DataFrame, table_name: str,
                         redshift_url: str = "jdbc:redshift://prod-rs:5439/warehouse") -> None:
        """Write data to AWS Redshift"""
        print(f"Writing to Redshift table: {table_name}")
        
        (df.write
         .format("jdbc")
         .option("url", redshift_url)
         .option("dbtable", table_name)
         .option("user", "etl_user")
         .option("password", "${REDSHIFT_PASSWORD}")
         .option("driver", "com.amazon.redshift.jdbc.Driver")
         .option("tempdir", "s3a://temp-bucket/redshift-temp")
         .mode("append")
         .save())

File: tests/mocks/scala/modular_app/DataOutputModule.scala
package com.company.etl.modules

import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
import org.apache.spark.sql.streaming.Trigger

/**
 * DataOutputModule - Handles all data writing operations
 * Writes to Hive, HDFS, JDBC, Kafka
 */
object DataOutputModule {
  
  /**
   * Write enriched transactions to Hive Gold layer
   */
  def writeEnrichedTransactions(df: DataFrame, env: String, runDate: String): Unit = {
    val tableName = s"${env}_gold.enriched_transactions"
    println(s"Writing enriched transactions to: $tableName")
    
    df.write
      .mode(SaveMode.Append)
      .partitionBy("run_date")
      .format("parquet")
      .saveAsTable(tableName)
  }
  
  /**
   * Write customer metrics to HDFS
   */
  def writeCustomerMetrics(df: DataFrame, basePath: String = "/data/gold", runDate: String): Unit = {
    val path = s"${basePath}/customer_metrics/date=${runDate}"
    println(s"Writing customer metrics to: $path")
    
    df.write
      .mode(SaveMode.Overwrite)
      .partitionBy("region")
      .parquet(path)
  }
  
  /**
   * Write fraud alerts to JDBC (PostgreSQL)
   */
  def writeFraudAlerts(df: DataFrame, jdbcUrl: String = "jdbc:postgresql://prod-pg:5432/alerts"): Unit = {
    println(s"Writing fraud alerts to: $jdbcUrl")
    
    df.write
      .mode(SaveMode.Append)
      .format("jdbc")
      .option("url", jdbcUrl)
      .option("dbtable", "fraud_alerts")
      .option("user", "etl_user")
      .option("password", sys.env.getOrElse("PG_PASSWORD", ""))
      .option("driver", "org.postgresql.Driver")
      .save()
  }
  
  /**
   * Write product summary to Hive reporting layer
   */
  def writeProductSummary(df: DataFrame, env: String): Unit = {
    val tableName = s"${env}_reports.product_daily_summary"
    println(s"Writing product summary to: $tableName")
    
    df.write
      .mode(SaveMode.Overwrite)
      .format("orc")
      .option("compression", "snappy")
      .saveAsTable(tableName)
  }
  
  /**
   * Write processed events to Kafka
   */
  def writeEventsToKafka(df: DataFrame, topic: String = "processed-events"): Unit = {
    println(s"Writing processed events to Kafka topic: $topic")
    
    df.selectExpr("CAST(customer_id AS STRING) AS key", "to_json(struct(*)) AS value")
      .writeStream
      .format("kafka")
      .option("kafka.bootstrap.servers", "kafka-broker:9092")
      .option("topic", topic)
      .option("checkpointLocation", s"/tmp/checkpoints/${topic}")
      .trigger(Trigger.ProcessingTime("30 seconds"))
      .start()
  }
  
  /**
   * Write aggregated data to Delta Lake
   */
  def writeToDeltalake(df: DataFrame, basePath: String = "/data/delta", tableName: String): Unit = {
    val path = s"${basePath}/${tableName}"
    println(s"Writing to Delta Lake: $path")
    
    df.write
      .format("delta")
      .mode(SaveMode.Overwrite)
      .option("mergeSchema", "true")
      .save(path)
  }
  
  /**
   * Export reports to S3 as CSV
   */
  def exportToS3(df: DataFrame, s3Bucket: String = "s3a://company-reports", reportName: String, runDate: String): Unit = {
    val path = s"${s3Bucket}/daily/${reportName}/date=${runDate}"
    println(s"Exporting report to S3: $path")
    
    df.write
      .mode(SaveMode.Overwrite)
      .option("header", "true")
      .option("compression", "gzip")
      .csv(path)
  }
}

